src/coreclr/jit/forwardsub.cpp

```diff
@@ -667,10 +661,23 @@ bool Compiler::fgForwardSubStatement(Statement* stmt)
     //
     // We may sometimes lose or change a type handle. Avoid substituting if so.
     //
-    if (gtGetStructHandleIfPresent(fwdSubNode) != gtGetStructHandleIfPresent(fsv.GetNode()))
+    // However, we allow free substitution of hardware SIMD types.
+    //
+    CORINFO_CLASS_HANDLE fwdHnd = gtGetStructHandleIfPresent(fwdSubNode);
+    CORINFO_CLASS_HANDLE useHnd = gtGetStructHandleIfPresent(fsv.GetNode());
+    if (fwdHnd != useHnd)
     {
-        JITDUMP(" would change struct handle (substitution)\n");
-        return false;
+        if ((fwdHnd == NO_CLASS_HANDLE) || (useHnd == NO_CLASS_HANDLE))
```

Comment:
Does this mean that if we have a `TYP_SIMD16` without a handle (such as because its a purely synthesized node) we can't substitute?

Reply:
Yes. Happy to just broadn this and check for simd type, rather than class handles, if that's correct.

Reply:
I'm not aware of anything that would impact this. We have logic for eliding `AsVector4` and `AsVector128` in import. Same for `Vector<T>` <-> `Vector128` or `Vector256` (which ever matches the size of `Vector<T>`)

This can basically be summed as:
* `TYP_SIMD8` can be subbed for another `TYP_SIMD8`
* `TYP_SIMD12` can be subbed for another `TYP_SIMD12`
* `TYP_SIMD16` can be subbed for another `TYP_SIMD16`
* `TYP_SIMD32` can be subbed for another `TYP_SIMD32`

If the `varType` doesn't match up, then substitution is invalid.

-----

I would generally expect we could do something like this

```cpp
if (varType1 != varType2)
{
    // mismatching varTypes, can't subsitute
    return false;
}

if (varTypeIsArithmetic(varType1) || varTypeIsSIMD(varType1))
{
    // We have a primitive arithmetic varType, like int, bool, short, etc (does TYP_INT vs TYP_UINT matter since we lose that sometimes?)
    // or one of the basic TYP_SIMD types and so substitution is possible and trivial
    return true;
}

// Now fallback to handle checking for other struct types, for byrefs, or other special considerations
```

Where for "well known types", we can just bypass the handle check entirely since they are primitives or SIMD. But for unknown struct types and other special scenarios, we need to do a handle check.

I'm unsure if integers are a problem for certain cases like where `TYP_UINT` is tracked on the stack as a `TYP_INT`.

Reply:
Of course, this is assuming there aren't any "gotchas" like @SingleAccretion was mentioning above.

Reply:
As below and above, the call arguments issue prevents us from being permissive here. We cannot lose precise handles for them if they're `varTypeIsStruct`.

Reply:
We don't always have handles though and we already lose some of that information in the case of the various `As*` casts?

Its not clear where or how this is working in such cases or if we have subtle bugs here (in which case they've probably been around for a few releases).

Reply:
> Its not clear where or how this is working in such cases

Currently, `impNormStructVal` takes care of preserving the handle and making sure it survives until morph. I know of no cases where it fails to do so.

Reply:
Do you mean preserving the handle in all the cases its required?

We have plenty of cases where `impHWIntrinsic` or `impSIMDIntrinsic` (the latter going back to 2014) will change the base type and so a handle simply won't be resolvable for the given node.

Reply:
> Do you mean preserving the handle in all the cases its required?

Correct.

Only two cases require precise handles today:
1) Call arguments.
2) Multu-reg returns. These are not an issue because they're handles rather specially and pessimized accordingly.

Reply:
So that means we'd need to special case forward substitutions when the use is a call (must always check handle), but otherwise it "should" be fine?

And once we fix morph to use the arg info directly, then we could change it (potentially) to be something more like what I was suggesting where we only compare the handles for unknown struct types?

Reply:
The end goal would be to allow forward sub even with mismatching class handles and leave the handling up to call args morphing. This would also get us down a path where we can remove `impNormStructVal` in `impPopCallArgs`. It was part of what I wanted to do with the call args work, and hope to continue after .NET 7 code complete.

Reply:
That sounds dangerous/complicated. I wouldn't expect something like `struct { byte x; }` to be subtitutable with say a `struct { int x; }`.

I'm guessing there is something I'm missing here...

Reply:
> So that means we'd need to special case forward substitutions when the use is a call (must always check handle), but otherwise it "should" be fine?

Yes, we could restrict this pessimization to calls/multi-reg returns only.

> And once we fix morph to use the arg info directly, then we could change it (potentially) to be something more like what I was suggesting where we only compare the handles for unknown struct types?

Yes.

Reply:
If you look at forward sub this handle check is just one of many cases where it won't substitute even if it should be legal and profitable (and likewise the legality and profitability analyses themselves can be improved).

Would be great if we could chip away at all these. I likely won't have much time to do this anytime soon.

Reply:
> I wouldn't expect something like `struct { byte x; }` to be subtitutable with say a `struct { int x; }`.

The ultimate goal is to make `ClassLayout::AreCompatible` the "type identity" for structs, and that would only allow same-sized structs.

Reply:
Right, yes, "mismatching class handles" was a bit too liberally phrased.